R/pkg_metric_has_news.R

```diff
@@ -0,0 +1,99 @@
+#' Assess a package for the presence of a NEWS file
+#'
+#' @param x a packge reference object
+#' @param ... additional arguments unused
+#'
+#' @return a \code{pkg_metric} object
+#' @export
+#'
+assess_has_news <- function(x, ...) {
+  UseMethod("assess_has_news")
+}
+
+
+
+#' @rdname assess_has_news
+#' @export
+#' @importFrom xml2 xml_find_all xml_attrs
+assess_has_news.pkg_remote <- function(x, ...) {
+  # scrape CRAN html for NEWS links
+  NEWS_links <- xml2::xml_find_all(x$web_html, xpath = '//a[.="NEWS"]')
+
+  # add NEWS link url metadata to package environment
+  x$materials_urls <- sprintf("%s/%s",
+    x$web_url,
+    vapply(xml2::xml_attrs(NEWS_links), "[", character(1L), "href"))
+
+  news_result <- rep(list(content = NULL, valid = TRUE), length(x$materials_urls))
+  structure(
+    news_result,
+    class = c("pkg_metric_has_news", "pkg_metric", class(news_result)))
+}
+
+
+
+#' @rdname assess_has_news
+#' @export
+assess_has_news.pkg_install <- function(x, ...) {
+  news_result <- assess_news_in_dir(system.file(package = x$name), ...)
+  structure(
+    news_result,
+    class = c("pkg_metric_has_news", "pkg_metric", class(news_result)))
+}
+
+
+
+#' @rdname assess_has_news
+#' @export
+assess_has_news.pkg_source <- function(x, ...) {
+  news_result <- assess_news_in_dir(x$path, ...)
+  structure(
+    news_result,
+    class = c("pkg_metric_has_news", "pkg_metric", class(news_result)))
+}
+
+
+
+#' @importFrom tools file_ext
+assess_news_in_dir <- function(x) {
+  news <- list()
+
+  # accommodate NEWS.Rd, NEWS.md, etc
+  news$files <- list.files(x, pattern = "^NEWS\\.*")
+  if (!length(news$files)) {
+    news$message <- "no NEWS file found"
+    return(news)
+  }
+
+  news$content <- rep(NULL, length(news$files))
+  news$valid <- vector(length(news$files), mode = "logical")
+
+  # attempt to parse all NEWS.* files
+  file_paths <- file.path(x, news$files)
+  for (i in seq_along(file_paths)) {
+    f <- file_paths[[i]]
+    tryCatch({
+      if (tolower(tools::file_ext(f)) == "rd") {
+        news$content[[i]] <- .tools()$.news_reader_default(f)
+      } else if (tolower(tools::file_ext(f)) == "md") {
+        # NOTE: should we do validation of markdown format?
+        news$content[[i]] <- readLines(f)
+      }
+      news$valid[[i]] <- TRUE
+    }, error = function(e) {
+      news$valid[[i]] <- FALSE
+    })
+  }
+
+  # NOTE: should we test whether NEWS file is up-to-date with latest version?
```

Comment:
That would be great if we could handle it easily. Authors may have different way to write the version into the NEWS. (or we identify a pattern like `[0-9].[0-9].[0-9]` to match with the DESCRIPTION part?) 

As long as they have a NEWS file, I would consider the author is responsive.

Reply:
Would it be okay to leave this for a future update? I'd prefer not to go down the rabbit hole of trying to validate all the different formats of NEWS files (`NEWS.md`, `NEWS.Rd`, `NEWS.html` (for remotely accessed NEWS files)) quite yet. 

Even those various text formats don't have any consistent syntax that is enforced across all packages and I'd be afraid that we could potentially invalidate otherwise good packages by being too prescriptive about what constitutes being "up-to-date".

Reply:
Yeah, I agree we can leave this for future. And we should keep the program neat and simple.